components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,240 @@
+			attrs, ok := cfg.ConvertedAttributes.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(attrs, cfg.ConvertedAttributes)
+			}
+			return NewRTSPCamera(ctx, attrs, logger)
+		},
+	})
+
+	config.RegisterComponentAttributeMapConverter(
+		camera.Subtype,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Attrs
+			attrs, err := config.TransformAttributeMapToStruct(&conf, attributes)
+			if err != nil {
+				return nil, err
+			}
+			result, ok := attrs.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(result, attrs)
+			}
+			return result, nil
+		},
+		&Attrs{},
+	)
+}
+
+// Attrs are the config attributes for an RTSP camera model.
+type Attrs struct {
+	Address          string                             `json:"rtsp_address"`
+	IntrinsicParams  *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters,omitempty"`
+	DistortionParams *transform.BrownConrady            `json:"distortion_parameters,omitempty"`
+}
+
+// Validate checks to see if the attributes of the model are valid.
+func (at *Attrs) Validate() error {
+	if prefix := strings.HasPrefix(at.Address, "rtsp://"); !prefix {
+		return errors.New(`rtsp_address must begin with "rtsp://"`)
+	}
+	if err := at.IntrinsicParams.CheckValid(); err != nil {
+		return err
+	}
+	if err := at.DistortionParams.CheckValid(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	u                       *url.URL
+	client                  *gortsplib.Client
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	activeBackgroundWorkers sync.WaitGroup
+	gotFirstFrameOnce       bool
+	latestFrame             atomic.Value
+	logger                  golog.Logger
+}
+
+// Close closes the camera. It always returns nil, but because of Close() interface, it needs to return an error.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	rc.cancelFunc()
+	rc.activeBackgroundWorkers.Wait()
+	if err := rc.client.Close(); err != nil && !errors.Is(err, clientTerminated) {
+		rc.logger.Infow("error while closing rtsp client:", "error", err)
+	}
+	return nil
+}
+
+// clientReconnectBackgroundWorker checks every 5 sec to see if the client is connected to the server, and reconnects if not.
+func (rc *rtspCamera) clientReconnectBackgroundWorker() {
+	rc.activeBackgroundWorkers.Add(1)
+	ticker := time.NewTicker(5 * time.Second)
+	goutils.ManagedGo(func() {
+		for {
+			if ok := goutils.SelectContextOrWaitChan(rc.cancelCtx, ticker.C); ok {
+				// use an OPTIONS request to see if the server is still responding to requests
+				res, err := rc.client.Options(rc.u)
+				if err != nil && (errors.Is(err, clientTerminated) ||
+					errors.Is(err, io.EOF) ||
```

Comment:
Is there something wrong with trying to `reconnectClient` on any errors? Are they some we definitely want to avoid?

Reply:
Some errors are really dumb - like "Oh, you made an OPTIONS request, but you are already in the PLAY state" which returns an error. So that error I would want to ignore. Or any errors regarding the OPTIONS themselves - this is just a way to check connectivity